### PR TITLE
 Target Allocator implementation (Part 2 - Target Allocator Image logic)

### PR DESCRIPTION
**Pull request by alexperez52**
### Description
Follows the changes in https://github.com/open-telemetry/opentelemetry-operator/pull/351

This PR addresses issue [open-telemetry/wg-prometheus#60.](https://github.com/open-telemetry/wg-prometheus/issues/60) This is part 2 of a 3 part update in which enhancements are presented to the OTel Operator to support the deployment of a Target Allocator. This update contains the logic for the Operator Managed target allocator image which will be used to perform target sharding for the Prometheus scrape targets and expose them to the collector(s) for consumption using http sd config.

The Design document for this part 2 can be found here: 
[Target Allocator Design Document](https://drive.google.com/file/d/1K5-8KaZLKIf1EAq8552vCZPFHVWmiNWt/view)

### Type of change

New feature 

### Testing
Unit tests were added to test the least connection algorithm flow as well as updating and removing targets.
More testing details can be found on the doc mentioned above.

cc: @Raul9595 @alolita

### CODE REVIEW

**Comment by alexperez52** [comment]
@dashpole @Aneurysm9 @alolita

**Comment by alolita** [comment]
@jpkrohling @Aneurysm9 can you please review. Look forward to any feedback you may have. ty!

**Comment by jpkrohling** [comment]
I'm planning on review this by the end of this week.

File: cmd/otel-allocator/main.go
Comment (Aneurysm9) [inline cmd/otel-allocator/main.go:114]: Does anything clean up these pod watchers when the config changes and `newAllocator()` is invoked?

File: cmd/otel-allocator/discovery/discovery.go
Comment (Raul9595) [reply cmd/otel-allocator/discovery/discovery.go:57]: Yes I totally agree with you. I just noticed an issue in the Collector repo open-telemetry/opentelemetry-collector#1867 which mentioned the exact same problem. We'll try to apply the same fix and get this resolved.

File: cmd/otel-allocator/main.go
Comment (Raul9595) [reply cmd/otel-allocator/main.go:116]: This has been handled in the latest update.

File: cmd/otel-allocator/discovery/discovery.go
Comment (Raul9595) [reply cmd/otel-allocator/discovery/discovery.go]: This has been handled in the latest update.

**Comment by jpkrohling** [comment]
Similar to the other PR: I left a couple of comments in the design doc. Let me know if I should go ahead and review this, or if you want to settle the open points first.

**Review by Aneurysm9 — DISMISSED** [review]

**Comment by alolita** [comment]
@alexperez52 @Raul9595 can you rerun the tests to make sure the E2E tests are passing.

File: cmd/otel-allocator/allocation/allocator.go
Comment (Raul9595) [reply cmd/otel-allocator/allocation/allocator.go:40]: If `SetCollectors` is called with 0 collectors, it would not enter the loop itself since _allocator.collectors_ will not have any elements thus ensuring `nextCollector` is never reached.

File: cmd/otel-allocator/allocation/allocator.go
Comment (dashpole) [reply cmd/otel-allocator/allocation/allocator.go:84]: nit: it is more idiomatic to have a `defer allocator.m.Unlock()` immediately after the Lock call.  This applies here and elsewhere.

File: cmd/otel-allocator/allocation/http.go
Comment (dashpole) [inline cmd/otel-allocator/allocation/http.go:84]: why do we need to iterate over the map if it is already indexed by collector and job name?

File: cmd/otel-allocator/allocation/http.go
Comment (dashpole) [inline cmd/otel-allocator/allocation/http.go:41]: Suggestion: A large portion of the code in here is finding things within lists of things that match an id (e.g. find the targets with job ID foo.  Readability could probably be improved by helper functions, such as `getCollectorTarget(collector, job)`, or `getAllTargets(job)`

**Review by dashpole — APPROVED** [review]
lgtm overall.  Mostly suggestions to make the code easier to read.

File: cmd/otel-allocator/allocation/http.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/http.go:84]: This is done to group back the targets back into their original groups that were split when introduced (if they landed in the same collector).
By just accessing the map we get something like:
`{`
`{Targets: [0.0.0.0:1000],`
` Labels: LabelCopy1},`
` {Targets: [0.0.0.0:2000],`
  `Labels: LabelCopy1}`
`}`
Instead we wanted (If they belonged to the same group before the split):
`{`
`{Targets: [0.0.0.0:1000, 0.0.0.0:2000],`
` Labels: LabelCopy1},`
`}`

File: cmd/otel-allocator/allocation/http.go
Comment (dashpole) [inline cmd/otel-allocator/allocation/http.go:118]: nit: this function doesn't need to be on Allocator, and doesn't need to pass in the request.  

Also, it might just be simpler to call `w.Header().Set("Content-Type", "application/json")` at the beginning of the TargetsHandler func, rather than in here, since all responses are json.  If you did that... i'm not sure you need this functino at all.

**Review by dashpole — APPROVED** [review]

File: cmd/otel-allocator/allocation/allocator.go
Comment (Aneurysm9) [inline cmd/otel-allocator/allocation/allocator.go:43]:
```suggestion
	var col *collector
	for _, v := range allocator.collectors {
		// If the initial collector is empty, set the initial collector to the first element of map
		if col == nil {
			col = v
```

Can we avoid the allocation of the initial `collector`?  Perhaps conditionally create an empty `collector` at return if the concern is about avoiding nil pointer checks everywhere the result is used.

**Review by Aneurysm9 — DISMISSED** [review]

File: cmd/otel-allocator/allocation/allocator.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator.go:43]: Yeah i agree. The initial purpose of this is to have an element to refer to as the 'min'

File: cmd/otel-allocator/allocation/http.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/http.go:118]: The purpose of the function was to be re-used in `TargetsHandler()` and `JobHandler()`

**Review by Aneurysm9 — DISMISSED** [review]

File: cmd/otel-allocator/collector/collector.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector.go:57]: Use the same logging facility as the other parts of the project.

File: cmd/otel-allocator/collector/collector.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector.go:82]: Split this into its own function, to make it easier to test.

File: cmd/otel-allocator/collector/collector.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector.go:85]: You need logging here and in the done, telling that the routine is closing.

File: cmd/otel-allocator/collector/collector_test.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector_test.go:16]: This needs a better name.

File: cmd/otel-allocator/collector/collector_test.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector_test.go:37]: This is probably an indication that your code needs a callback, to notify when it's ready. Waiting N seconds will make this test either inefficient, as it waits more than it needs, or intermittent when it takes longer than 1s for the client to be ready.

File: cmd/otel-allocator/collector/collector_test.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector_test.go:39]: Each one of the runs should be on its own test function, so that they can be executed individually. You might want to look at TestMain to prepare the environment that will be used for all tests.

File: cmd/otel-allocator/collector/collector_test.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector_test.go:44]: Where are you using this?

**Review by jpkrohling — COMMENTED** [review]
Sorry that I couldn't complete the review, but will make another pass later today or tomorrow.

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:42]: This test is only testing to make sure that the `initTargets` got initialized and sent to the `Allocator` struct. 
The `initTargets` acts as mock data for the incoming targets. These targets are processed and allocated among the collectors _evenly_ but this test is only testing that they were processed and no targets got dropped that weren't supposed to be dropped.

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:65]: Yes, they should be called together always if the desired result is to process incoming targets and then re-adjust the internal information to remove and add targets according to the `targetsWaiting`

File: cmd/otel-allocator/collector/collector_test.go
Comment (Raul9595) [reply cmd/otel-allocator/collector/collector_test.go]: This was added just to start with an one collector instance in the cluster. I have removed this as it is not required and it can be misleading as we dont have the functionality of adding a pod.

File: cmd/otel-allocator/collector/collector_test.go
Comment (Raul9595) [reply cmd/otel-allocator/collector/collector_test.go]: Yes I agree. I have replaced it with sync.WaitGroup which is a better way to test I feel.

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:171]: Its impossible to tell which collector will have a certain amount of targets since they are being added and removed in the multiple steps. This test is just checking that we go back to a ground up allocation when collectors change. 
Perhaps i should remove the noisy steps and just test for the functionality mentioned

File: cmd/otel-allocator/allocation/http.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/http.go:12]: I'm a bit confused as to how this could be done

**Comment by jpkrohling** [comment]
I'll review this again soon, but please, try to get a PR fixing the intermittent tests that the part 1 introduced.

**Comment by alexperez52** [comment]
> I'll review this again soon, but please, try to get a PR fixing the intermittent tests that the part 1 introduced.

Seems like the PR merged today solved the issue with the e2e tests

File: cmd/otel-allocator/go.mod
Comment (jpkrohling) [inline cmd/otel-allocator/go.mod:8]: Is there anything specific that you need from this?

File: cmd/otel-allocator/go.mod
Comment (jpkrohling) [inline cmd/otel-allocator/go.mod:13]: This probably needs to be updated as well.

File: cmd/otel-allocator/main.go
Comment (jpkrohling) [inline cmd/otel-allocator/main.go:16]: Remove this empty line.

File: cmd/otel-allocator/main.go
Comment (jpkrohling) [inline cmd/otel-allocator/main.go:66]: Have you tested this logic in a container deployed on Kubernetes?

File: cmd/otel-allocator/main.go
Comment (jpkrohling) [inline cmd/otel-allocator/main.go:106]: Your `server` struct could actually provide two HTTP handler functions, like:

```go
func (*server) JobHandler(w http.ResponseWriter, r *http.Request)
```

This way, you can just pass `s.JobHandler` as a parameter to the router.

File: cmd/otel-allocator/collector/collector.go
Comment (Raul9595) [reply cmd/otel-allocator/collector/collector.go:119]: These are being added in the test function. This is only used in the tests.

File: cmd/otel-allocator/collector/collector_test.go
Comment (Raul9595) [reply cmd/otel-allocator/collector/collector_test.go:56]: Yes this is to ensure that the test function waits for the go routine to execute a set number of times.

File: cmd/otel-allocator/discovery/discovery.go
Comment (Raul9595) [reply cmd/otel-allocator/discovery/discovery.go:21]: The discovery manager requires a go-kit/log and currently does not allow any other logger.

File: cmd/otel-allocator/main.go
Comment (Raul9595) [reply cmd/otel-allocator/main.go:66]: Yes this has been tested on a minikube cluster locally.

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:105]: `TestAddingAndRemovingTargets` tests that targets are successfully added with the expected with the expected _key_
`TestCollectorBalanceWhenAddingTargets` tests the the balance on an easily divisible number for the purpose of _even_ allocation
`TestCollectorBalanceWhenAddingAndRemovingAtRandom` uses a more _random_ approach where targets are both added and removed while verifying that at each step there isnt above a 15% difference in workload from collector to collector

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:171]: The current implementation will do a re allocation from scratch when collector information is modified so this test is for the purposes of verifying that targets were re allocated from scratch. Using a divisible ratio of targets to collectors, its verified that the targets are allocated evenly.

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/allocator_test.go:193]: With this current implementation no. `SetCollectors` would always be followed by a Reallocation

File: cmd/otel-allocator/allocation/http.go
Comment (alexperez52) [reply cmd/otel-allocator/allocation/http.go:34]: This makes sense. I'll update it in the next commit :-)

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (jpkrohling) [reply cmd/otel-allocator/allocation/allocator_test.go:105]: Not sure I get the value-add between the second and the third, but assuming there's enough value justifying the added complexity, isn't it covering the third case covering the second? Meaning, do we need the second? And the first?

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (jpkrohling) [reply cmd/otel-allocator/allocation/allocator_test.go:171]: Sorry, I should have been clearer, but do we still need both tests? Isn't one of them covering what the other is doing?

File: cmd/otel-allocator/allocation/allocator_test.go
Comment (jpkrohling) [reply cmd/otel-allocator/allocation/allocator_test.go:193]: I would then automatically do a rebalance whenever there's a change to the data that affects the balancing. Like, when the list of targets changes, rebalance. When the list of collectors changes, rebalance again.

File: cmd/otel-allocator/collector/collector.go
Comment (jpkrohling) [reply cmd/otel-allocator/collector/collector.go:119]: How is this then working outside of tests? If there's no wg added, Done will cause a panic here... This are you sure there was a manual test with this version of the code?

**Comment by jpkrohling** [comment]
Ping me when this is ready for review again.

File: cmd/otel-allocator/collector/collector.go
Comment (Raul9595) [reply cmd/otel-allocator/collector/collector.go]: Sorry I meant that the old version was tested. Not with this version. I have updated it to handle the panic. `runWatch` will only be called during testing now.

File: cmd/otel-allocator/collector/collector.go
Comment (jpkrohling) [inline cmd/otel-allocator/collector/collector.go:88]: This is not really the way to go here... You might want to have a channel here, or perhaps accept a callback function.

**Comment by jpkrohling** [comment]
The tests are failing and should have been fixed by #410. I tried to rebase this PR to get that change but got a merge conflict. Please, rebase this PR.

**Comment by jpkrohling** [comment]
I understand that @Aneurysm9 will be maintaining this code, so, I'll wait for his review before doing a final round myself.

**Comment by alolita** [comment]
Hi @jpkrohling thanks for taking a look. 

@Aneurysm9 can you do a final review and let's get this merged. Ty!!!

**Comment by alolita** [comment]
Ping @Aneurysm9 please review so that @jpkrohling can merge. Thx.

**Review by Aneurysm9 — APPROVED** [review]
There are some things I'd like to refactor, particularly in the `allocation` package and the `main()` function, but I think this is a solid starting point and provides a usable, and useful, service.

@jpkrohling if you'd like to see that refactoring done before merging this I'd understand, but I also think it could be good to land it as it is to make the capability available and see what happens when people attempt to put it to use.

**Comment by jpkrohling** [comment]
The e2e tests for Kubernetes 1.19 failed and I just restarted. I'll merge this as soon as the tests pass.

**Review by jpkrohling — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**
